jobserver/authorization/utils.py

```diff
@@ -1,11 +1,102 @@
 import itertools
 
+from .converters import org_converter, project_converter
 
-def has_permission(user, permission):
+
+def _validate_context(allowed_keys, context):
+    """
+    Ensure keys and types-of-values in context are correct
+
+    When passing in context to either has_permission or has_role we want to
+    help developers track down incorrect keys or incorrect values.
+
+    For example, given this call:
+
+        _validate_context(["project"], {"projects": project})
+
+    we want to tell the develoer that "projects" is an invalid key.
+
+    Similarly we want to make sure the contextual objects passed in are valid
+    for the keys used, take the following function call:
+
+        _validate_context(["project"], {"project": user})
+
+    We want to tell the developer that a "project" key must map to a Project
+    (model) instance.
+    """
+    if not context:
+        return
+
+    # checks based on the allowed list passed in
+    unknown_keys = set(context.keys()) - set(allowed_keys)
+    if unknown_keys:
+        msg = "Some invalid keys were used in the context:"
+
+        for key in unknown_keys:
+            msg += f"\n - {key}"
+
+        raise ValueError(msg)
+
+    # import here to avoid circular imports
+    from jobserver import models
+
+    type_map = {
+        "org": models.Org,
+        "project": models.Project,
+    }
+    incorrect_types = {}
+    for key, value in context.items():
+        model = type_map[key]
+        if not isinstance(value, model):
+            incorrect_types[key] = type(value)
+
+    if not incorrect_types:
+        return
+
+    msg = "Some invalid values were used in the context:"
+    for key, value in incorrect_types.items():
+        expected = type_map[key]
+        msg += (
+            f'\n - key "{key}" got a {value.__name__}, expected a {expected.__name__}'
+        )
+
+    raise ValueError(msg)
```

Comment:
I can see how this all works, but it feels like quite a lot of work.  If instead you had `has_permission(user, permission)`, `has_permission_on_project(user, permission, project)`, and `has_permission_on_org(user, permission, org)`, I think you'd end up with simpler code.

Reply:
A core goal of the OFP was to have minimal entry point functions (ideally one!) to the framework, this change would add O(N) entry points, one for each Object where local roles can be applied (Project, Org, etc) and one for each Object where global roles can be applied (User, and we're expecting Groups at some point too).

It also means a developer has to know where permissions are coming from to know which functions to call since Roles can exist in multiple places.  Take, for example George, a member of both OrgA and ProjectA.  ProjectA is part of OrgA.  A caller of `has_permission` can pass in both OrgA and ProjectA as context for the lookup, without having to know if they should look at the Org or Project specifically.

Reply:
You've clearly thought about this much more than I have, so happy to defer.

But I might have misunderstood something.  I'd assumed that when checking whether a user can do something to a project, you'd always pass that project in the context.  But are you saying that you can also pass that project's organisation in the context instead?  (Part of my problem is that I'm not looking at any real examples of how this is used.)

(Also, a general point: I'm not sure whether it's right to think about building a framework -- a framework has to be useful for lots of users across lots of applications, while here we're building a single application, with a small number of users.  As such, we don't have to offer something as general-purpose as a framework might have to.)